tests/clients/test_fenda_ditto_client.py

```diff
@@ -0,0 +1,171 @@
+
+
+@pytest.mark.parametrize("type,model", [(FendaDittoClient, FendaModel(FeatureCnn(), FeatureCnn(), FendaHeadCnn()))])
+def test_forming_ditto_loss(get_client: FendaDittoClient) -> None:  # noqa
+    torch.manual_seed(42)
+    fenda_ditto_client = get_client
+    fenda_ditto_client.global_model = SequentiallySplitExchangeBaseModel(FeatureCnn(), HeadCnn())
+    fenda_ditto_client.parameter_exchanger = FullParameterExchanger()
+    fenda_ditto_client.lam = 1.0
+    config: Config = {"current_server_round": 2}
+    fenda_ditto_client.model.first_feature_extractor.load_state_dict(
+        fenda_ditto_client.global_model.base_module.state_dict()
+    )  # feature extracor is given to FENDA model
+
+    params = [val.cpu().numpy() + 0.1 for _, val in fenda_ditto_client.global_model.state_dict().items()]
+    fenda_ditto_client.set_parameters(params, config, fitting_round=True)
+    fenda_ditto_client.update_before_train(4)
+
+    ditto_loss = fenda_ditto_client.ditto_drift_loss_function(
+        fenda_ditto_client.model.first_feature_extractor,
+        fenda_ditto_client.initial_global_tensors,
+        fenda_ditto_client.lam,
+    )
+
+    assert fenda_ditto_client.lam == 1.0
+    assert pytest.approx(ditto_loss.detach().item(), abs=0.02) == (fenda_ditto_client.lam / 2.0) * (
+        1.5 + 0.06 + 24 + 0.16
+    )
+
+
+@pytest.mark.parametrize("type,model", [(FendaDittoClient, FendaModel(FeatureCnn(), FeatureCnn(), FendaHeadCnn()))])
+def test_compute_loss(get_client: FendaDittoClient) -> None:  # noqa
+    torch.manual_seed(42)
+    fenda_ditto_client = get_client
+    fenda_ditto_client.global_model = SequentiallySplitExchangeBaseModel(FeatureCnn(), HeadCnn())
+    fenda_ditto_client.parameter_exchanger = FullParameterExchanger()
+    config: Config = {"current_server_round": 2}
+    fenda_ditto_client.criterion = torch.nn.CrossEntropyLoss()
+    fenda_ditto_client.lam = 1.0
+
+    params = [val.cpu().numpy() for _, val in fenda_ditto_client.global_model.state_dict().items()]
+    fenda_ditto_client.set_parameters(params, config, fitting_round=True)
+    fenda_ditto_client.update_before_train(4)
+
+    # Training: both FENDA models are updated
+    state_dict = fenda_ditto_client.global_model.base_module.state_dict()
+    for key in state_dict:
+        state_dict[key] += 0.1
+    fenda_ditto_client.model.first_feature_extractor.load_state_dict(state_dict)
+    fenda_ditto_client.model.second_feature_extractor.load_state_dict(state_dict)
+
+    preds = {"global": torch.tensor([[1.0, 0.0], [0.0, 1.0]]), "local": torch.tensor([[1.0, 0.0], [0.0, 1.0]])}
+    target = torch.tensor([[1.0, 0.0], [1.0, 0.0]])
+    training_loss = fenda_ditto_client.compute_training_loss(preds, {}, target)
+    fenda_ditto_client.global_model.eval()
+    fenda_ditto_client.model.eval()
+    evaluation_loss = fenda_ditto_client.compute_evaluation_loss(preds, {}, target)
+    print(evaluation_loss)
```

Comment:
Dropping print statements in tests is good practice 🙂

Reply:
removed!